added batch method generator

letusfly85 opened the pull request:
if there is some problem, please tell me :drop:

seratch commented on scalikejdbc-mapper-generator-core/src/main/scala/scalikejdbc/mapper/CodeGenerator.scala:
```diff
@@ -376,6 +376,34 @@ class CodeGenerator(table: Table, specifiedClassName: Option[String] = None)(imp
+     *   Seq(
+     *     'id -> entity.id,
+     *     'name -> entity.name,
+     *     'birthday -> entity.birthday)
```
compilation error. `'birthday -> entity.birthday))`?

seratch commented on scalikejdbc-mapper-generator-core/src/main/scala/scalikejdbc/mapper/CodeGenerator.scala:
```diff
@@ -376,6 +376,34 @@ class CodeGenerator(table: Table, specifiedClassName: Option[String] = None)(imp
+     * }
+     * }}}
+     */
+    val batchMethod = {
```
We should skip this if the `id` columns is an auto-increment one. Check `table.autoIncrementColumns.size == 1` as same as `createMethod`.

seratch replied:
Please rename this. How about batchInsertMethod?

seratch commented on scalikejdbc-mapper-generator-core/src/main/scala/scalikejdbc/mapper/CodeGenerator.scala:
```diff
@@ -376,6 +376,34 @@ class CodeGenerator(table: Table, specifiedClassName: Option[String] = None)(imp
+     */
+    val batchMethod = {
+      // def batch=(
+      1.indent + s"def batch(list: List[" + className + "])(implicit session: DBSession = autoSession) {" + eol +
```
def batchInsert

seratch replied:
For historical reason, findAll returns List value but I think accepting Seq[Entity] here is reasonable enough.

seratch replied:
What do you think about this? >I think accepting Seq[Entity] here is reasonable enough.

seratch commented:
Please squash commits into one commit after fixing my comments.

xuwei-k commented:
generated code

```scala
def batchInsert(list: List[Programmer])(implicit session: DBSession = autoSession) {
  val params: Seq[Seq[(Symbol, Any)]] = list.map(elem => 
    Seq(
      'name -> elem.name,
      't1 -> elem.t1,
      't2 -> elem.t2,
      't3 -> elem.t3,
      '`type` -> elem.`type`))
      SQL("""insert into programmers values (
      {name},
      {t1},
      {t2},
      {t3},
      {`type`}
    )""").batchByName(params: _*).apply()
  }
```

```
[info] [error] /tmp/sbt_264932d1/gen/src/main/scala/app/models/Programmer.scala:111: unclosed character literal
[info] [error]         '`type` -> elem.`type`))
[info] [error]         ^
[info] [error] /tmp/sbt_264932d1/gen/src/main/scala/app/models/Programmer.scala:111: unclosed quoted identifier
[info] [error]         '`type` -> elem.`type`))
[info] [error]                              ^
[info] [error] /tmp/sbt_264932d1/gen/src/main/scala/app/models/Programmer.scala:119: ')' expected but '}' found.
[info] [error]     }
[info] [error]     ^
[info] [error] /tmp/sbt_264932d1/gen/src/main/scala/app/models/Programmers.scala:111: unclosed character literal
[info] [error]         '`type` -> elem.`type`))
[info] [error]         ^
[info] [error] /tmp/sbt_264932d1/gen/src/main/scala/app/models/Programmers.scala:111: unclosed quoted identifier
[info] [error]         '`type` -> elem.`type`))
[info] [error]                              ^
[info] [error] /tmp/sbt_264932d1/gen/src/main/scala/app/models/Programmers.scala:119: ')' expected but '}' found.
[info] [error]     }
[info] [error]     ^
[info] [error] 6 errors found
[info] [error] (compile:compile) Compilation failed
```

https://travis-ci.org/scalikejdbc/scalikejdbc/jobs/69830808#L2275

letusfly85 commented:
Sorry, I checked about `type` column
And I also see a scalikejdbc-test package.

xuwei-k commented:
> I also see a scalikejdbc-test package.

generator test codes is here

https://github.com/scalikejdbc/scalikejdbc/tree/2.2.7/scalikejdbc-mapper-generator/src/sbt-test/scalikejdbc-mapper-generator

 (not [scalikejdbc-test ](https://github.com/scalikejdbc/scalikejdbc/tree/2.2.7/scalikejdbc-test) )

xuwei-k commented on scalikejdbc-mapper-generator-core/src/main/scala/scalikejdbc/mapper/CodeGenerator.scala:
```diff
@@ -376,6 +376,45 @@ class CodeGenerator(table: Table, specifiedClassName: Option[String] = None)(imp
+     *   Seq(
+     *     'id -> elem.id,
+     *     'name -> elem.name,
+     *     'birthday -> elem.birthday)
```
maybe missing closing parenthesis

letusfly85 commented:
Thank you!
I'll check it!

letusfly85 commented:
I modified comment of batchInsertMethod, and added test code.

seratch commented:
Please verify your `git config user.email` to use your GitHub account's email address. https://help.github.com/articles/setting-your-email-in-git/

We'd love to list you as a contributor when merging this PR. https://github.com/scalikejdbc/scalikejdbc/graphs/contributors

letusfly85 commented:
>Please verify your git config user.email to use your GitHub account's email address.
Thank you, I'll configure and push again :sweat_drops:

xuwei-k commented on scalikejdbc-mapper-generator-core/src/main/scala/scalikejdbc/mapper/CodeGenerator.scala:
```diff
@@ -376,6 +376,49 @@ class CodeGenerator(table: Table, specifiedClassName: Option[String] = None)(imp
+          allColumns
+
+      // def batchInsert=(
+      1.indent + s"def batchInsert(list: List[" + className + "])(implicit session: DBSession = autoSession) {" + eol +
```
```scala
(implicit session: DBSession = autoSession)
```

should be

```scala
(implicit session: DBSession$defaultAutoSession)
```


https://github.com/scalikejdbc/scalikejdbc/blob/2.2.7/scalikejdbc-mapper-generator-core/src/main/scala/scalikejdbc/mapper/CodeGenerator.scala#L200

xuwei-k commented:
> I modified comment of batchInsertMethod

it seems still not modified

xuwei-k commented:
`SQLBatch#apply` returns `Seq[Int]`

https://github.com/scalikejdbc/scalikejdbc/blob/2.2.7/scalikejdbc-core/src/main/scala/scalikejdbc/SQL.scala#L444

I think `batchInsert` returns `Seq[Int]` too (not `Unit`)

letusfly85 commented:
@xuwei-k 

Thank you.

I modified paren of Seq[Int],

also changed defaultAutosession by seeing above and [line 333](https://github.com/scalikejdbc/scalikejdbc/blob/2.2.7/scalikejdbc-mapper-generator-core/src/main/scala/scalikejdbc/mapper/CodeGenerator.scala#L333),

and changed return value to Seq[Int]

Thank you.

xuwei-k commented:
https://travis-ci.org/scalikejdbc/scalikejdbc/jobs/69856499#L2548

```
[info] [info] ProgrammersTest:
[info] [info] Programmer
[info] [info] - should be available *** FAILED ***
[info] [info]   List(6) did not equal 3 (ProgrammersTest.scala:46)
[info] [info] ScalaTest
[info] [info] Run completed in 2 seconds, 82 milliseconds.
[info] [info] Total number of tests run: 1
[info] [info] Suites: completed 1, aborted 0
[info] [info] Tests: succeeded 0, failed 1, canceled 0, ignored 0, pending 0
[info] [info] *** 1 TEST FAILED ***
[info] [error] Failed: Total 1, Failed 1, Errors 0, Passed 0
[info] [error] Failed tests:
[info] [error] 	app.models.ProgrammersTest
[info] [error] (test:testOnly) sbt.TestsFailedException: Tests unsuccessful
[info] [error] Total time: 3 s, completed Jul 7, 2015 9:25:22 AM
[error] x scalikejdbc-mapper-generator / gen
[error]    {line 22}  Command failed: + failed
```

seratch commented:
Please add test code generation for batchInsert too. https://github.com/letusfly85/scalikejdbc/blob/batch_method2generator/scalikejdbc-mapper-generator-core/src/main/scala/scalikejdbc/mapper/CodeGenerator.scala#L788

letusfly85 commented:
Thank you.
I added batchInsert test template.

seratch commented on scalikejdbc-mapper-generator-core/src/main/scala/scalikejdbc/mapper/CodeGenerator.scala:
```diff
@@ -878,6 +934,7 @@ class CodeGenerator(table: Table, specifiedClassName: Option[String] = None)(imp
           |    "create new record"            ! autoRollback().create ^
           |    "save a record"                ! autoRollback().save ^
           |    "destroy a record"             ! autoRollback().destroy ^
+          |    "batch insert records"         ! autoRollback().batchInsert ^
```
How about `"perform batch insert"` instead?

seratch commented on scalikejdbc-mapper-generator-core/src/main/scala/scalikejdbc/mapper/CodeGenerator.scala:
```diff
@@ -854,6 +904,12 @@ class CodeGenerator(table: Table, specifiedClassName: Option[String] = None)(imp
           |    }
+          |    "batch insert records" in new AutoRollback {
+          |      val list = %className%.findAll()
+          |      list.foreach(%className%.destroy)
```
`%className%.findAll().foreach(%className%.destroy())` would be simple.

seratch commented on scalikejdbc-mapper-generator-core/src/main/scala/scalikejdbc/mapper/CodeGenerator.scala:
```diff
@@ -376,6 +376,49 @@ class CodeGenerator(table: Table, specifiedClassName: Option[String] = None)(imp
+          allColumns
+
+      // def batchInsert=(
+      1.indent + s"def batchInsert(list: List[" + className + "])(implicit session: DBSession" + defaultAutoSession + "): Seq[Int] = {" + eol +
```
https://github.com/scalikejdbc/scalikejdbc/pull/427#discussion_r34099832

letusfly85 commented:
Thanks, I modified to 'perform batch insert'

xuwei-k commented on scalikejdbc-mapper-generator-core/src/main/scala/scalikejdbc/mapper/CodeGenerator.scala:
```diff
@@ -376,6 +376,49 @@ class CodeGenerator(table: Table, specifiedClassName: Option[String] = None)(imp
+     * }}}
+     */
+    val batchInsertMethod = {
+      val autoIncrement = table.autoIncrementColumns.size == 1
```
Maybe `table.autoIncrementColumns.nonEmpty` ?

hmm, but there is same codes in `val createMethod`

seratch commented on scalikejdbc-mapper-generator-core/src/main/scala/scalikejdbc/mapper/CodeGenerator.scala:
```diff
@@ -854,6 +904,12 @@ class CodeGenerator(table: Table, specifiedClassName: Option[String] = None)(imp
           |      shouldBeNone.isDefined should beFalse
           |    }
+          |    "perform batch insert" in new AutoRollback {
+          |      val list = %className%.findAll()
```
https://github.com/scalikejdbc/scalikejdbc/pull/427#discussion_r34116054

letusfly85 commented:
@xuwei-k 

>Maybe table.autoIncrementColumns.nonEmpty ?

I felt natural to see 'table.autoIncrementColumns.size == 1'
because there is a limit that a table cannot have multi auto columns...

letusfly85 commented:
@xuwei-k 

Sorry, I want to change my opinion above.

I saw below and thought today.

* http://docs.oracle.com/javase/8/docs/api/java/sql/DatabaseMetaData.html
* http://dev.mysql.com/doc/refman/5.5/en/example-auto-increment.html

I thought that
it is obvious that table can have only one auto_increment column (it is a rule),
however
there is no method in JDBC such as 'table.hasAutoIncrementColumn: Boolean'.

And below , I want to know about table having some auto_increment column, but don't need to know about the columns.size is 1 or 0.

```scala
val autoIncrement = table.autoIncrementColumns.size == 1
```

So, I think it is better  to change from 'table.autoIncrementColumns.size' to 'table.autoIncrementColumns. nonEmpty' on both createMethod and batchInsertMethod.

Should I post a new issue or PR? :sweat_drops:

seratch commented:
Although I think this topic is not so important, `table.autoIncrementColumns.nonEmpty` looks fine here. Please fix createMethod too.

seratch commented:
@letusfly85 Take a look at my comments too.

letusfly85 commented:
@seratch @xuwei-k 

I see, thank you!

letusfly85 commented:
I changed.

```scala
%className%.findAll().foreach(%className%.destroy)
```

```scala
table.autoIncrementColumns.nonEmpty
```

seratch commented on scalikejdbc-mapper-generator-core/src/main/scala/scalikejdbc/mapper/CodeGenerator.scala:
```diff
@@ -924,6 +981,12 @@ class CodeGenerator(table: Table, specifiedClassName: Option[String] = None)(imp
           |      shouldBeNone.isDefined should beFalse
           |    }
+          |    def batchInsert = this {
+          |      val list = %className%.findAll()
```
@letusfly85 Sorry, I've misunderstood your code >%className%.findAll().foreach(%className%.destroy)

Please fix like this:

```scala
val entities = %className%.findAll()
entities.foreach(e => %className%.destroy(e))
val batchInserted = %className%.batchInsert(entities)
batchInserted.size should be_>(0)
```

letusfly85 replied:
No problem, I see.

And if 'entities' is good, I want to also change from 'list' to 'entities' in 'batchInsertMethod'.

seratch replied:
Please avoid using `list`.

letusfly85 replied:
I see. Thank you.

letusfly85 commented:
I did some modifying.

And, below failure doesn't have any relation of my modify, I think..

https://travis-ci.org/scalikejdbc/scalikejdbc/jobs/70145947#L2748

I tried to push to check again.

letusfly85 commented:
Was the travisCI condition?

I passed CI test.

letusfly85 commented:
Sorry, I want to check about camel case column.

letusfly85 commented:
I finished modifying, if you have some point, please let me know..

seratch commented on scalikejdbc-mapper-generator-core/src/main/scala/scalikejdbc/mapper/CodeGenerator.scala:
```diff
@@ -854,6 +904,12 @@ class CodeGenerator(table: Table, specifiedClassName: Option[String] = None)(imp
+          |    "perform batch insert" in new AutoRollback {
+          |      val entities = %className%.findAll()
+          |      entities.foreach(e => %className%.destroy(e))
+          |      val batchInserted = %className%.batchInsert(entities.toList)
```
toList is unnecessary.

letusfly85 replied:
If I remove toList, the below test case will fail..

https://github.com/scalikejdbc/scalikejdbc/blob/master/scalikejdbc-mapper-generator/src/sbt-test/scalikejdbc-mapper-generator/gen/test#L77

letusfly85 commented:
I removed toList,
However I used Seq for argument.

seratch commented on scalikejdbc-mapper-generator/src/sbt-test/scalikejdbc-mapper-generator/gen/src/test/scala/ProgrammersTest.scala:
```diff
@@ -38,6 +38,9 @@ class ProgrammersTest extends fixture.FlatSpec with Matchers with AutoRollback {
     Programmers.findAll().toSet should equal(Set.empty[Programmers])
     Programmers.countAll() should equal(0)
+
+    val res = Programmers.batchInsert(programmers.toList)
```
toList is not necessary here

letusfly85 replied:
@seratch 

if I remove toList, it will fails..
toSeq or toList are reasonable here I think. 

around on line 18, `val programmers` start as Set by using to 'toSet' method 
```scala
val programmers = List(Some("aaa"), Some("bbb"), None).map(name =>
  Programmers.create(name = name, t1 = new DateTime(2014, 12, 31, 20, 0))
     ).toSet	
```

line 18's `toSet` is necessary to tset findAll method for non order collection. (@xuwei-k told me , and I missed it..)
So, if I want to use this `programmers: Set[Programmers]` for batchInsert, I have to change it toSeq or toList on line 42...

letusfly85 replied:
@seratch

To remove `toList` I changed, around on line 18's, from 'Set' to 'Seq' of `val programmers`,
and applied `toSet` for results of `findAll` method, also removed `toList` from a Random.shuffle argment `programmers`.

seratch commented:
Thanks. As I suggested [here](used Seq for argument), that's fine for me. >use Seq instead

letusfly85 commented:
Thank you for replying, I also removed toList from ProgrammersTest.scala.

xuwei-k commented on scalikejdbc-mapper-generator/src/sbt-test/scalikejdbc-mapper-generator/gen/src/test/scala/ProgrammersTest.scala:
```diff
@@ -17,27 +17,30 @@ class ProgrammersTest extends fixture.FlatSpec with Matchers with AutoRollback {
     val invalidId = Int.MinValue
     Programmers.find(invalidId) should equal(None)
-    Programmers.findAll().toSet should equal(programmers)
+    Programmers.findAll() should equal(programmers)
```
Why `.toSet` removed? `Programmers#findAll` does not specify `ORDER BY`.

letusfly85 replied:
Sorry, I missed, I will restore.

letusfly85 commented:
I restore at once and reply to note, please see it.

seratch commented:
Thank you for your contribution :+1:

letusfly85 commented:
Thank you, too :smile: !!!
